### ISSUE TITLE
Add support for sharing state across analyzer instances

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#6324]
Currently, our diagnostic analyzer API provides no support for sharing any state between diagnostic analyzers. This causes extreme pain or performance issues for our customers when they have some expensive computation and/or large data to share across analyzer instances. For example:

  1. Group of analyzers that need identical configuration settings from an additional file - they have identical computation logic and resulting data, and would get a real performance benefit if they can share this data across analyzer instances without doing any complicated lifetime management.

  2. StyleCop analyzers want to walk through all trees in a compilation and identify generated files - they want to do it just once per compilation across all analyzer instances. In absence of any API support, they ended up creating a static conditional weak table keyed on compilation, which caused GC related perf problems when executing in the IDE. See https://github.com/DotNetAnalyzers/StyleCopAnalyzers/pull/1671 for the workaround they put to avoid CWT.

An ideal solution here is to provide an analyzer API to share per-compilation and per-additional file data across analyzer instances, so that analyzer authors don't have to explicitly deal with lifetime management for such state. I am going to propose an initial API for this feature, and we can iterate over the best design for it.

**Principles:**
1. API should guarantee easy state access/creation within all analyzer callbacks.
2. Analyzer author needs to handle concurrent access/update to shared state instance. (@JohnHamby proposes we should think about enhancing the API to have this as an opt-in rather a requirement)
3. [Implementation] No conditional weak tables for storing share state: CWT are evil in most instances, the shared state should be strongly referenced within CompilationWithAnalyzers/AnalyzerDriver instance.

**Proposed API (per-compilation state):**
Add the below API to each [DiagnosticAnalysisContext](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs) parameter type for analyzer callbacks:

```C#
TState GetOrCreateSharedState<T, TState>(Key<T> key, Func<Key<T>, TState> createSharedState) where T: object, TState: object
```

where `public sealed class Key<T> { }` is a separate definition.

**Comment by mavasani** [comment]
Tagging people who might be interested in this feature/design discussion: @sharwell @tmeschter @srivatsn @JohnHamby @heejaechang @pharring

**Comment by JohnHamby** [comment]
Can we do better in terms of the safety of the concurrency model?

**Comment by JohnHamby** [comment]
If we are going to require analyzers that share state to be in the same assembly, can't any analyzers that want to share state be written as a single analyzer with multiple actions?

**Comment by JohnHamby** [comment]
Adding @gafter @mattwar

**Comment by sharwell** [comment]
> Can't any analyzers that want to share state be written as a single analyzer with multiple actions?

We would prefer to avoid that, as it complicates the implementation of single analyzers. For example, it's currently easy to review the behavior of [SA1018 (NullableTypeSymbolsMustNotBePrecededBySpace)](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1018NullableTypeSymbolsMustNotBePrecededBySpace.cs) against the [documentation for that rule](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1018.md). It would not be nearly so easy if the analysis behavior for 150+ other violations was included in the same analyzer.

**Comment by sharwell** [comment]
:memo: Another thing to consider here. In DotNetAnalyzers/StyleCopAnalyzers#1683 and DotNetAnalyzers/StyleCopAnalyzers#1689, we *dramatically* reduced the number of allocations required to register our analyzers with a new compilation. However, due to our use of shared state we are currently unable to remove the [closure created here](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/1.0.0-beta015/StyleCop.Analyzers/StyleCop.Analyzers/AnalyzerExtensions.cs#L129-L141), which results in an allocation for every call we make to `RegisterSyntaxNodeAction` (there are perhaps 1000 such calls overall).

An ideal solution to state management would allow us to reduce or eliminate these per-compilation allocations.

**Comment by heejaechang** [comment]
@mavasani why limiting sharing state per assembly make threading model more safe? I thought we only serialize analyzer execution not assembly? are we saying if an analyzer from a assembly register shared state, we will serialize all analyzer executions from that assembly regardless whether a particular analyzer from the assembly uses shared state or not?

**Comment by sharwell** [comment]
> Allow state sharing only between analyzer instances from the same assembly - this is needed to ensure a secure threat model, we don't want unrelated analyzers to be able to access/manipulate each other's state.

In my opinion, we don't need this goal. The intended outcome has questionable value at best, but the impact could result in very real limitations for analyzer developers. For example, in the future we may want to provide a common API for detecting generated code in a StyleCop-compliant manner, which many analyzers then use.

> API should guarantee easy state access/manipulation within all analyzer callbacks. However, analyzer author needs to handle concurrent access/update to shared state instance.

We've found that simply having a lazily-initialized value associated with a compilation would be sufficient. For example, all we need for generated files support is to associate a single (initially empty) instance of `ConcurrentDictionary<SyntaxTree, bool>` with a `Compilation` instance, and then access that dictionary in each analyzer. We never need to modify the dictionary *instance* which is associated with the compilation, even though we do have code to modify the *contents* of that instance.

I believe the priorities for the API should be:

1. Allow lazily- or singly-initialized object to be associated with a compilation (allocating one dictionary even eagerly would not be a problem)
2. Reduce the number of allocations required when registering analyzer callbacks for the case where shared state is used

> [Implementation] No conditional weak tables for storing share state: CWT are evil in most instances, the shared state should be strongly referenced within CompilationWithAnalyzers/AnalyzerDriver instance.

More specifically, after the data is initialized we need access to this data to be lock-free. The implementation of CWT does not allow us to meet this requirement.

**Comment by heejaechang** [comment]
@mavasani ah, I misinterpret the meaning of thread model. sorry. ignore my previous comment.

**Comment by sharwell** [comment]
@heejaechang 

> why limiting sharing state per assembly make threading model more safe?

The original post says *threat*, which I interpreted as a discussion of security, as opposed to *thread* which would be a discussion of concurrency.

**Comment by heejaechang** [comment]
@mavasani so this reduce repeated work per each analyzer but analyzer need to repeat the same work per compilation, right? I think that is good goal. we should clear the shared state as soon as all analysis of a compilation has finished. otherwise we could get into memory leak.

**Comment by mavasani** [comment]
@heejaechang @sharwell We need some kind of threat model/security to ensure unrelated analyzers cannot view/modify state or worse throw exceptions based on the assumption about the data type of the state. I was proposing we start with drawing a boundary around the assembly in which analyzers exists, and can extend the API later to support sharing data across assemblies. However, we do need some measures to prevent unrelated analyzers from accidentally/purposefully corrupt shared data.

**Comment by heejaechang** [comment]
@mavasani @sharwell 2 questions.

1. I think we can change "key" from string to object. by doing it, the author can decide whether they want to share their state or not by making the object key shared between analyzers. since we don't restrict the state as a serializable object, I think we can just use object as key.
 
2. why limit TState as object if we are not going to use CWT?

**Comment by sharwell** [comment]
> so this reduce repeated work per each analyzer but analyzer need to repeat the same work per compilation, right?

We've actually been successful in accomplishing this, but it's been a challenge with many pitfalls. We've completely changed the way we manage per-compilation state twice as a result of performance problems that snuck in. The primary benefit of a common API for us will be the elimination of many additional allocations (several hundred at least) which occur during the registration of our analyzers.

> We need some kind of threat model/security to ensure unrelated analyzers cannot view/modify state or worse throw exceptions based on the assumption about the data type of the state.

If you change the key to be a `Type` which is compared by reference (as opposed to a `string`), then the simple use of a type which is not publicly exposed as a key would be sufficient to address this concern.

**Comment by pharring** [comment]
If you use "object" as the key instead of "string", then you can leave it up to the analyzer authors to determine "sharability". If you need private state, then use a "private static readonly object" key. If you need something visible to other analyzers in the same assembly, then use "internal", etc.

#5542 had that

**Comment by mavasani** [comment]
@pharring that is surely a better approach, let me update the original comment accordingly.

**Comment by heejaechang** [comment]
@mavasani can we remove this API.
bool TryGetSharedState<TState>(string key, out TState value) where TState: object

can one always use GetOrCreate ? that will remove people to assume there is any dependency we guarantee.

**Comment by heejaechang** [comment]
@mavasani also, can we remove "State" from the API and rather put "Cache" in them. so that people doesn't use it as a state but use as a "Cache". implication is, "state" can break functionality but cache can't. this also remove assumption on dependency and also, we don't need to re-think our statefull and stateless analyzer story.

**Comment by mavasani** [comment]
> can we remove this API.
 bool TryGetSharedState(string key, out TState value) where TState: object

Yes, probably but API design probably want to decide that. If the analyzer wants to have `null` as an allowable value for the data for exceptional circumstances, then a reader invoking GetOrCreate needs to do an additional null check. Just a minor API preference..

> can we remove "State" from the API and rather put "Cache" in them. so that people doesn't use it as a state but use as a "Cache". 
I thought we explicitly want it to be strongly referenced state, that lives the lifetime of the compilation. We don't want to create conditional weak table or weakly referenced property bags or any kind of cache, that has already proven to be bad for IDE perf. Am I misunderstanding your comment?

**Comment by gafter** [comment]
I would prefer this API be typesafe, for example as in  #5708.

**Comment by heejaechang** [comment]
@mavasani 

"Yes, probably but API design probably want to decide that. If the analyzer wants to have  null  as an allowable value for the data for exceptional circumstances, then a reader invoking GetOrCreate needs to do an additional null check. Just a minor API preference.."

user can still return null from GetOrCreateXXX if they want to.

"I thought we explicitly want it to be strongly referenced state, that lives the lifetime of the compilation. We don't want to create conditional weak table or weakly referenced property bags or any kind of cache, that has already proven to be bad for IDE perf. Am I misunderstanding your comment?"

this is not what I meant. implementation-wise, leave it exactly as you suggested.

what I meant was just remove "State" from API and rather put "Catch" in them. and tell them you can't depends on the data for your functional correctness. so stateless or statefull analyzer both still can use this API without making them stateful analyzer. conjunction with the removing of "TryXXXX" API, all user of the API, which uses GetOrCreateXXX should be able to create the cache if it doesn't exist.

**Comment by sharwell** [comment]
I like the simplicity of a single `GetOrCreateXXX` method, and it resolves every single case we have where we need to register per-compilation analyzer callbacks as opposed to calling back to static methods which are registered once for the lifetime of the analyzer assembly.

For this to work, it is essential that the fast path (case where the item already exists) be very efficient in a concurrent environment, because it could be invoked in each of the analysis callbacks (potentially dozens of times per `SyntaxNode` depending on the number of analyzers enabled).

> I would prefer this API be typesafe, for example as in #5708.

We could get this by

**Comment by mavasani** [comment]
@gafter I have updated the proposed API to take a `TKey` instead of `object` key, hopefully this is what you meant.

@heejaechang 
> tell them you can't depends on the data for your functional correctness
Why would we do so? The fact that you want per-compilation shared state makes you a stateful analyzer, and we should be able to guarantee functional correctness by keeping this object alive as long as the underlying compilation lives. We can just store the state on CompilationWithAnalyzers instance, which has strong reference to the underlying compilation on which analyzers are executed. Analyzers can safely assume that we will initialize this state only once per compilation.

**Comment by gafter** [comment]
@mavasani No, it doesn't enforce any relationship between the type of the key and the type of the value. The implementation of `TryGetSharedState` will contain a cast that can fail. Please see #5708 for a model of type safety for this kind of API.

**Comment by sharwell** [comment]
>  I have updated the proposed API to take a `TKey` instead of `object` key, hopefully this is what you meant.

Define a class `Key<T>` as follows:

```csharp
public sealed class Key<T> { }
```

Then update the API to require the key's type match the created value:

```csharp
TState GetOrCreateSharedState<TState>(Key<TState> key, Func<object, TState> createSharedState)
    where TState: object
```

**Comment by gafter** [comment]
@sharwell Yes, like that.

**Comment by heejaechang** [comment]
@mavasani first analyzer become stateful analyzer. you get one benefit and you lost one - concurrency since state basically implies dependency.

since it is state that they relies on for functionality, the order of state mutation must be correct and observable from all callers.

making it not state is much simpler and stateless analyzer can use it as well.

**Comment by sharwell** [comment]
> Yes, almost like that. get rid of the object:

I copied from the original comment above. I would actually prefer the constraint be removed.

**Comment by sharwell** [comment]
> making it not state is much simpler and stateless analyzer can use it as well.

How exactly do you determine whether an analyzer is stateful or stateless?

Also, if the only operation available is `GetOrCreateXXX`, then there is no negative impact on concurrency. As long as the factory method passed in is always the same for a given key, there will be no way to determine within the analyzer whether the value was created or existed previously.

**Comment by heejaechang** [comment]
@sharwell 

var state = GetOrCreateSharedState(myKey, k => new ConcurrentDictionary<ISymbol, bool>())
... in analyzer1
state.Add(methodSymbol, fFoundReference);
... in analyzer2
var state = ....
state.TryGetValue(methodSymbol, out fdoWeHaveReference)

now two has dependency, we need to make sure analyzer1 runs first and then analyzer2.

if it is cache that can't used for functionality, one can't write the code above.

**Comment by heejaechang** [comment]
@sharwell if it is strictly a cache, one will write something like this
.. in analyzer 1
var state = GetOrCreateSharedCache(myKey, k => new GeneratedFileCache);
var generatedFile = state.GetOrAdd(file1, fGeneratedFile);
if (generatedFile) ...

.. in analyzer2
var state = ...
var generatedFile = state.GetOrAdd(file1, fGeneratedFile);
if (generatedFile) ...

no dependency. if one doesn't exist, it will just simply re-calculate. now since it doesn't affect state-ness of an analyzer, any analyzer can use this API.

**Comment by sharwell** [comment]
@heejaechang It is my understanding that the relative order of execution of two analyzers is already non-deterministic. The proposed API wouldn't change that. Also, without the ability to express the relative ordering of analyzer 1 and 2 in your example, I don't think analyzer developers would be assuming that code like you showed would work.

**Comment by heejaechang** [comment]
@sharwell  then, what do you mean by shared state?

**Comment by paul1956** [comment]
I need to read the contents of an additional file once, parse it and store the result (in my case a list) somewhere Analyzers can access it. However when the file changes and the Analyzers rerun I need to reparse the file, similar to how ASP.net allows a cache to have the ability to track changes in a file and invalidate the cache. In addition to being able to cache/perist a list of generated files.

**Comment by heejaechang** [comment]
@paul1956 in Roslyn, compilation is an immutable snapshot, so file change doesn't matter for a compilation. and since the proposed API cache the information only for a particular compilation, you will re-calculate your cache on every new compilation anyway, so it shouldn't matter.

if you want to cache your info across multiple compilations, you probably need to create your own mechanism since engine doesn't support that. especially since Roslyn support in memory forking, just caching latest result might not always work.

**Comment by paul1956** [comment]
@heejaechang The list of Generated Files changes very rarely and changes to Analyzer options by the user, change ever less frequently (maybe only time in a projects life) but they are among the most expensive operations because they involve opening and parsing a text file. Any solution should address these factors.

**Comment by heejaechang** [comment]
@paul1956 I think what you are asking is having a way to get delta between 2 compilations. I don't think there is a way for it right now in any way. at least in compiler layer since compiler doesn't care about 2 different snapshot.

**Comment by paul1956** [comment]
@heejaechang I don't think so. Today I am told if my options file (saved as an AdditionalFile) changes all the Analyzers run again (Wonderful part of my issue is solved). Now I need that file change to invalidate the saved state being discussed above, I am even willing to provide a list of files I care about (the option file) like I do for a traditional Cache. The bottom line is Roslyn needs to solve the overhead of IsGeneratedFile (by providing a cached public implementation) and getting Cached/parsed User Options to an Analyzer. Reading multiple files, that almost never change, per keystroke does not make a lot of sense. I leave the implementation up to people a lot smarter than me. For both of these issues I see as many different solutions as there are Open Source Roslyn based projects and none of them are perfect yet.

**Comment by sharwell** [comment]
> The bottom line is Roslyn needs to solve the overhead of IsGeneratedFile (by providing a cached public implementation) and getting Cached/parsed User Options to an Analyzer.

Our profiling to date indicates this is not the case. For all practical purposes, the two items we use a cache for (calculating whether a document is generated and loading the settings file) don't even appear on profiling reports.

**Comment by paul1956** [comment]
I guess we all need to copy your (cache) implementation since this functionality is not provided by Roslyn, which was really my point.

**Comment by tmeschter** [comment]
My preference would be to allow analyzers to associated state with additional files specifically, in addition to compilations as a whole. This would be analagous to how `SyntaxTree`s are associated with source files, rather than being tied to compilations, and would also allow the state to be reused across compilations--it only needs to be recomputed when the underlying `SourceText` changes.

However, this wouldn't address example number 2, so we would still need a way to associate state with the compilation as a whole.

**Comment by mavasani** [comment]
I completely agree with Tom. We are talking about 2 different scenarios here, each of which seem to be valuable:
1. Share per-compilation state between analyzers - changes with each compilation snapshot. I proposed an API for this scenario in the original description.
2. Share state across compilations for the lifetime of any of the analyzer options (currently analyzer options just supports additional files) - we need to think about an API for this.

@JohnHamby @srivatsn - Let us talk about both these scenarios at the next design meeting and upload the resulting proposal here, sounds reasonable?

**Comment by heejaechang** [comment]
@mavasani let the lifetime of user sate to go over compilation is dangerous thing. if that is the case, user can use static field themselves already.

**Comment by heejaechang** [comment]
I am not sure why additional file is special, the problem @paul1956 currently have might require some cache only for additional files, but in general, it looks like, what he wants is he want to minimize work he need to do by knowing what has changed since last time (when cache is built) it has run. currently we don't give or maintain that information since we treat each compilation as a new fresh start.

the original proposal was keeping that principle, but share information between analyzers in that same snapshot. now discussion is moved to sharing information between compilations. for that scenario to be any useful generally, there must be a way for user to figure out what has changed between two compilations (base and current) to figure out what information is still valid.

**Comment by mavasani** [comment]
@heejaechang Your point is true in general when the lifetime of state is not associated with anything. Here we have a concrete scenario for analyzer authors where lifetime of state needs to be associated with an additional file snapshot - whose content can be identical across compilations. Proposal here is that this is an important enough scenario to expose an API to allow storing or caching data across compilations based on additional file snapshot. The author just needs to be aware that such data can live across multiple compilations, and hence live longer in memory. They need to be judicious on what to store versus recompute for each compilation. 

The guiding principle here is - for any entity which represents an analysis input (compilation, additional files), analyzer author should be able to store custom data associated with the lifetime of the entity. The diagnostic engine controls the lifetime of all entities (compilation at compiler layer, additional files at workspace layer) and is the best component to manage lifetime of the associated data - any attempt by the analyzer author to manage lifetime with static fields is likely to cause memory leaks.

**Comment by heejaechang** [comment]
@mavasani then why not for source text and syntax tree and other entity?

**Comment by heejaechang** [comment]
@mavasani if we really want to have such property bag, then we should force it to be serializable. if you see any entity that lives across compilations, all those are recoverable (serializable) and we do not let it alive in memory.

**Comment by paul1956** [comment]
Static doesn't work because the Analyzer has no way to know the AdditionalFiles changed without the overhead of at least comparing time stamps. Which is something I might do until a better solution is provided.

**Comment by JohnHamby** [comment]
The message that we need to do something here is clear. I think the issues requiring solution are also clear. The Roslyn team is going to need to do some thinking and offline discussion, and will come back with a more complete proposal in a few days.

**Comment by JohnHamby** [comment]
Here's a proposal to cover reading an external resource from multiple analyzers. It does not cover sharing mutable state between analyzers.

A significant aspect of the issues discussed here is a need for multiple analyzers to read configuration, options, or directive context from external resources. Reading and parsing a resource can be expensive, motivating a desire to read a given resource once per compilation or per analysis session rather than once per analyzer per compilation. Here is a proposed API with the following goals:
  -- Minimize the number of times a resource is read and parsed.
  -- Guarantee that a resource will be reread if it changes between accesses.

To be added to:
  -- AnalysisContext
  -- CompilationStartAnalysisContext
  -- CompilationAnalysisContext

We propose:
  bool TryGetResource<T>(Key<T> key, SourceText text, Func<SourceText, T> reader, out T result)

Semantics:
  -- At first invocation for a given key value and a given text value, invoke the reader function with the text value. If the reader returns normally, set result to the value returned by the reader and return true. If the reader terminates with an exception, set result to Default(T) and return false.
  -- A subsequent invocation with an equivalent key value and equivalent text value is permitted (and is likely) to bypass invoking the reader, set result to the previous result, and return the same value as the prior invocation. However, it is also permitted to invoke the reader function again.
  -- A subsequent invocation with a different key value or a different text value must invoke the reader.

Caching can cross contexts. For example, if there is a call to AnalysisContext.TryGetResource in one analyzer and an equivalent call to CompilationStartAnalysisContext.TryGetResource in another, the resource will likely be read only once. The lifetime of a cached result is arbitrary, but is expected to be as long as practical without introducing excessive memory pressure.

Expectations on clients:
  -- Every invocation of a given reader with equivalent text will produce equivalent results.
  -- All invocations of these methods for a given key will supply readers with equivalent behavior and text values intended to represent (versions of) the same resource.
  -- The value returned by a reader will not be mutated carelessly.
  -- For a given key, a single analyzer will make no more than one invocation of AnalysisContext.TryGetResource and no more than one invocation per compilation of CompilationStartAnalysisContext.TryGetResource or CompilationAnalysisContext.TryGetResource, and will invoke no more than one of these methods.

Violating these expectations can produce difficult-to-explain behavioral inconsistencies.

**Comment by sharwell** [comment]
> For a given key, a single analyzer will make no more than one invocation of AnalysisContext.TryGetResource and no more than one invocation per compilation of CompilationStartAnalysisContext.TryGetResource or CompilationAnalysisContext.TryGetResource, and will invoke no more than one of these methods.

I don't understand the need for either of these limitations. Can you clarify how this benefits the implementation and the reason it needs to be different from the behavior described by your other expectations?

**Comment by JohnHamby** [comment]
Violating these restrictions might produce a surprise. If the resource changes between invocations, the result you get back will also change. If the resource has not changed and you supply an equivalent reader delegate, you might get back an identical object from two invocations, or you might get different (but equivalent) results. If you supply a different delegate to two invocations with the same key, the second invocation might return the results of the first and not call the delegate supplied with the second invocation.

The implementation won't enforce these restrictions. It also won't do anything to guarantee predictable behavior if they are violated.

**Comment by sharwell** [comment]
> If the resource changes between invocations, the result you get back will also change. If the resource has not changed and you supply an equivalent reader delegate, you might get back an identical object from two invocations, or you might get different (but equivalent) results. If you supply a different delegate to two invocations with the same key, the second invocation might return the results of the first and not call the delegate supplied with the second invocation.

Based on the description in the previous comment, none of these behaviors is unexpected. I think what you currently have documented as an expectation on clients would be better worded as a recommendation that care be taken, along with an example like you provided in the follow-up. With the expectation in place, clients will not be able to leverage this functionality even if the developers understand the special considerations and know that they would not negatively impact their analyzers.

**Comment by mavasani** [comment]
Thanks @JohnHamby for the proposal.
@sharwell @paul1956 Apart from the minor modifications to the wording of the semantics/expectations, does the proposal seem adequate for your requirements?

**Comment by sharwell** [comment]
> Apart from the minor modifications to the wording of the semantics/expectations, does the proposal seem adequate for your requirements?

Overall, yes. Some questions remain which may or may not be in scope of this:

1. Where does the `SourceText` instance come from (second parameter of proposed API)?
2. Would a similar API be added for code fix contexts?

**Comment by mavasani** [comment]
> Where does the `SourceText` instance come from (second parameter of proposed API)?

Additional files: http://source.roslyn.io/Microsoft.CodeAnalysis/R/b33eb1a9beed263a.html
Regular source files: http://source.roslyn.io/Microsoft.CodeAnalysis/R/5967ebe19d531e5d.html

> Would a similar API be added for code fix contexts?

Seems reasonable to me. I'll file a separate issue for it.
